Thread wakeup visualization

Lectem opened the pull request:
This PR does the following things:
- Allow to lock the highlighted thread using left mouse click or lane popup menu
- Add more descriptions for wait reasons in View::DecodeContextSwitchReason
- Stores the Wakeup CPU so that we may display it in the CPU data view
- `QueueThreadWakeup` now contains cpu and adjust info (currently unused). Fits in the padding, no struct size change.
- `QueueContextSwitch` now also contains thread priorities information (windows+linux) and CState (windows), but currently unused by the profiler. Fits in the padding, no struct size change.
- Fixes Linux `EventWakeup` which tries to remove PERF_SAMPLE_CALLCHAIN on the wrong member. Set the whole `perf_event_attr` explicitely instead.
- Fixes Windows ReadyThread event being dropped if triggered before the thread switches out (this is confirmed to be normal behaviour, probably due to thread being switched out but context switch actually happening later)
- Bumps tracy file version and protocol versions

Screenshots of the wakeup viz:

![image](https://github.com/user-attachments/assets/56272959-a56f-4f4f-9415-b434e1768704)
![image](https://github.com/user-attachments/assets/79ba1a10-fd1c-4a9b-852f-334c928630a1)

In the future we should be able to retrieve the readying thread information, such as its callstack, by iterating the per-core context switch data. Ideally we would also have a way to draw an arrow between the thread lanes and not just the CPU cores data ?

wolfpld commented on profiler/src/profiler/TracyTimelineItemThread.cpp:
```diff
@@ -144,7 +144,7 @@ int64_t TimelineItemThread::RangeEnd() const
 void TimelineItemThread::HeaderTooltip( const char* label ) const
 {
-    m_view.HighlightThread( m_thread->id );
+    m_view.HighlightThread(m_thread->id, false);
```
Please use the correct code style across the changes.

Lectem replied:
Done, sorry !

wolfpld commented on public/common/TracyVersion.hpp:
```diff
@@ -6,8 +6,8 @@ namespace tracy
 enum { Major = 0 };
-enum { Minor = 11 };
-enum { Patch = 2 };
+enum { Minor = 12 };
```
Keep minor at 11.

wolfpld commented:
Please update the manual to explain the changes you've made. As it now is I don't really know what the screenshots show. Having a proper writeup will help with code review.

Lectem commented:
I added a new paragraph and illustration to the manual (pages 78-79 for me).
Please tell me what you think about it.

wolfpld commented on manual/tracy.tex:
```diff
@@ -3542,9 +3542,45 @@ \subsubsection{Timeline view}
+When the \faMousePointer{}~mouse pointer is hovered over either the CPU data zone or the thread timeline label, Tracy will display a line connecting all zones associated with the selected thread. This can be used to quickly see how the thread migrated across the CPU cores. 
 
-Clicking the \LMB{}~left mouse button on a tracked thread will make it visible on the timeline if it was either hidden or collapsed before.
+It will also add lines starting with a filed circle to denote wake up events. Those are useful to pinpoint the origin of a thread waking up, especially when holding locks. It may also start from an empty region, denoting the time at which the kernel made chose to schedule or boost the priority of your thread. Wake ups will have a different color based on the reason for which the thread was waitin ...
```
> Wake ups will have a different color based on the reason for which the thread was waiting to be scheduled. 

Is there a way to know what these colors will be?

Lectem replied:
Right now not really, but the "wait reason" can be seen from the thread lanes. The idea behind coloring it was to quickly see the different kinds of wait reason, but I think in the long term we might want to categorize each wait reason and regroup them ?

wolfpld commented on manual/tracy.tex:
```diff
@@ -3542,9 +3542,45 @@ \subsubsection{Timeline view}
+When the \faMousePointer{}~mouse pointer is hovered over either the CPU data zone or the thread timeline label, Tracy will display a line connecting all zones associated with the selected thread. This can be used to quickly see how the thread migrated across the CPU cores. 
 
-Clicking the \LMB{}~left mouse button on a tracked thread will make it visible on the timeline if it was either hidden or collapsed before.
+It will also add lines starting with a filed circle to denote wake up events. Those are useful to pinpoint the origin of a thread waking up, especially when holding locks. It may also start from an empty region, denoting the time at which the kernel made chose to schedule or boost the priority of your thread. Wake ups will have a different color based on the reason for which the thread was waitin ...
```
> made chose to

made choice to

wolfpld commented on manual/tracy.tex:
```diff
@@ -3542,9 +3542,45 @@ \subsubsection{Timeline view}
+
+In the above picture, \emph{Thread B} migrates from \emph{Core 3} to \emph{Core 4} due to a wake up from \emph{Thread A}. Then it migrates from \emph{Core 4} to \emph{Core 1}.
+
+Clicking the \LMB{}~left mouse button on a tracked thread will make it visible on the timeline if it was either hidden or collapsed before. It will also lock the selected thread so that you may move explore data while retaining the visualization of thread migrations and wake up events.
```
> you may move explore data

pan explored

Lectem commented:
Fixed the latest comments (I think), and also added 

> Clicking again somewhere empty on the timeline with the left mouse button will unlock the selection.

wolfpld commented on profiler/src/profiler/TracyTimelineItemThread.cpp:
```diff
@@ -275,6 +275,14 @@ void TimelineItemThread::DrawOverlay( const ImVec2& ul, const ImVec2& dr )
 
+void TimelineItemThread::DrawExtraPopupItems()
+{
+    if( ImGui::MenuItem( ICON_FA_TIMELINE " Select in CPU timeline" ) )
```
Is this also shown when there is no CPU timeline?

Lectem replied:
Good point, I'll try to make it conditional

wolfpld commented on server/TracyEvent.hpp:
```diff
@@ -530,36 +530,95 @@ struct CrashEvent
+        Win32_WrIoRing          = 40,
+        Win32_WrMdlCache        = 41,
+        Win32_WrRcu             = 42,
+        Win32_MaximumWaitReason,
```
Where is this list sourced from?

Lectem replied:
This comes from the windows driver kit, though not documented online you can find the Rust versions here https://microsoft.github.io/windows-docs-rs/doc/windows/Wdk/System/SystemServices/constant.WrIoRing.html or from the Process Hacker docs https://ntdoc.m417z.com/kwait_reason

wolfpld commented on server/TracyEvent.hpp:
```diff
@@ -530,36 +530,95 @@ struct CrashEvent
-    tracy_force_inline void SetStartCpu( int64_t start, uint8_t cpu ) { assert( start < (int64_t)( 1ull << 47 ) ); _start_cpu = ( uint64_t( start ) << 16 ) | cpu; }
-    tracy_force_inline void SetEndReasonState( int64_t end, int8_t reason, int8_t state ) { assert( end < (int64_t)( 1ull << 47 ) ); _end_reason_state = ( uint64_t( end ) << 16 ) | ( uint64_t( reason ) << 8 ) | uint8_t( state ); }
+    tracy_force_inline void SetStartCpu( int64_t start, uint8_t cpu ) { assert( start < (int64_t)( 1ull << 47 ) ); _start.SetVal( start ); _cpu = cpu; }
+    tracy_force_inline void SetEndReasonState( int64_t end, int8_t reason, int8_t state ) { assert( end < (int64_t)( 1ull << 47 ) ); _end = end; _reason = reason; _state = state; }
```
These functions existed only to have the setting operation as a single store. With the way data is now split it makes no sense to keep these.

wolfpld commented on server/TracyWorker.cpp:
```diff
@@ -6772,9 +6781,12 @@ void Worker::ProcessContextSwitch( const QueueContextSwitch& ev )
             assert( item.End() == -1 );
+            //TODO: it may happen that events are being dropped (for example due to breaking in the debugger, or we are simply too slow to handle the events)
+            //      We should handle this properly in some way, but it is unclear how. We can't even really detect it properly other than when cpu doesn't match.
+            assert( item.Cpu() == ev.cpu );
```
Yeah, it will happen. This shouldn't be an assert, because some people will run with no asserts and it will be ignored anyways, and other people will complain that the profiler crashes.

Lectem replied:
Should I just remove the assert and ignore the issue or do we actually want to generate a new context switch?
It's not really clear how to fill the data if we want to generate a new cswitch. (what time do we use for end of the previous cswitch, and for start of the new one).

wolfpld replied:
It's all guesswork when data is missing. The capture should ideally be able to eventually recover proper state after more data is received.

Lectem replied:
I suppose we could detect dropped events from ETW display it somehow in the UI but it seems out of scope for this PR. Will leave the comment (with a bit more detail if needed) but remove the assert, is that OK with you? (ie, same behavior as before, but documented)

wolfpld replied:
It's ok.

wolfpld commented:
Lines on the CPU data graph no longer work as intended.

Correct behavior:

<img width="207" alt="obraz" src="https://github.com/user-attachments/assets/8f329ae4-3c56-4e2a-933d-25b173162983" />

After moving the leftmost element out of view:

<img width="276" alt="obraz" src="https://github.com/user-attachments/assets/606b43ac-819c-47d0-b741-2447b7a04f40" />

wolfpld commented:
Elements too small to be visible are not culled, which results in an unreadable display and severely degraded performance.

<img width="496" alt="obraz" src="https://github.com/user-attachments/assets/0a8f9b4f-c635-4923-88e6-52705d6835d8" />

wolfpld commented:
Selecting a thread on the CPU data list hijacks the highlight functionality. Threads hovered over with the mouse will not be highlighted until the selection is removed. These should be two separate things.

wolfpld commented:
I want to see how the two wakeup colors differ, but there is no explanation in the tooltip.

<img width="448" alt="obraz" src="https://github.com/user-attachments/assets/28c79100-09e8-46e2-8f98-8f2dc27dbc93" />

Lectem commented:
> Lines on the CPU data graph no longer work as intended.

Oversight on my part due to wrong using the worng signedness for the time. Fixed.

> The vertical lines are drawn on top of the above and below CPU lanes

This was intentional as I felt it looked better, but I changed it to fit the height of the core.

> Elements too small to be visible are not culled, which results in an unreadable display and severely degraded performance.

I am now only displaying those if the wait duration spans more than 30px or if the element would be > 0.5px. That way we hit a nice spot between usability and performance. I'm open to suggestions if you think there are better ways to do it.

> I want to see how the two wakeup colors differ, but there is no explanation in the tooltip.

I added information about the waits and wakeups in the tooltip

> Selecting a thread on the CPU data list hijacks the highlight functionality. Threads hovered over with the mouse will not be highlighted until the selection is removed. These should be two separate things.

This brings a few questions:
If I select a thread using the click on the timeline, then hover another (from either the cpu data timeline or the thread lane)
- Should I display both? 
- Should the hovered one have priority over the selected one?
- Should I fade out the selected one?

wolfpld commented:
> This brings a few questions: If I select a thread using the click on the timeline, then hover another (from either the cpu data timeline or the thread lane)
> 
>     * Should I display both?
> 
>     * Should the hovered one have priority over the selected one?
> 
>     * Should I fade out the selected one?

Display both, with different colors. See GPU zones or locks for reference.

wolfpld commented on profiler/src/profiler/TracyTimelineItemThread.cpp:
```diff
@@ -275,6 +275,14 @@ void TimelineItemThread::DrawOverlay( const ImVec2& ul, const ImVec2& dr )
 
+void TimelineItemThread::DrawExtraPopupItems()
+{
+    if( m_view.GetViewData().drawCpuUsageGraph && ImGui::MenuItem( ICON_FA_TIMELINE " Select in CPU timeline" ) )
```
That's not correct. You need to check if the cpu usage data exists. Looking at the queries the info window makes is a good starting point.

wolfpld commented:
Unselecting doesn't work properly.
- Clicking in a window (not on the timeline) will unselect the thread.
- Unselecting only works if CPU data is drawn. It's enough to collapse the CPU data section to not be able to unselect the thread.

wolfpld commented:
The marker has incorrect offset.

<img width="100" alt="image" src="https://github.com/user-attachments/assets/2e70f2c2-49ea-43ba-8b21-8f6ffe9c34a3" />

wolfpld commented:
The wakeup line does not adapt to DPI.

<img width="96" alt="image" src="https://github.com/user-attachments/assets/6a27f50d-aec9-4b35-ab9c-c915b67a0915" />

wolfpld commented:
Is the data gathered correctly on Linux?

I have worker threads that are sleeping on a condition variable, and the job dispatching thread is waking the workers. But all I see is that the worker threads are "woken by kernel".

<img width="291" alt="image" src="https://github.com/user-attachments/assets/36a12951-9f06-4188-bb87-ea2b0ea635c6" />

What I see instead are some strange wakeup attributions that don't make much sense.

<img width="611" alt="image" src="https://github.com/user-attachments/assets/da480a1b-98a1-42b7-a358-aed658bf1a97" />

Worker#16 is seemingly woken by Worker#12, but the worker threads do not communicate with each other. Workers may block each other on access to the job queue (as seen on the lock at the bottom of the image), but when Worker#16 obtains the lock, the wakeup reason is still "woken by kernel".

Lectem commented:
It seems that there are subtle differences between sched_wakeup and sched_waking (there's also sched_wakeup_new) and that the origin CPU might not be the one of the wakee depending on various things ...

See https://perfetto.dev/docs/data-sources/cpu-scheduling

I'll investigate a bit more.

Note that on windows, it is possible that a thread seemingly unrelated readies (wakes up) another if it triggers the scheduler by calling any kind of synchronization or syscall. (this mostly happens when thread was already ready to be scheduled). From what I saw on windows, we're now displaying the same things as WPA and Pix.

Lectem commented:
> The marker has incorrect offset.
> <img alt="image" width="100" src="https://private-user-images.githubusercontent.com/600573/432041248-2e70f2c2-49ea-43ba-8b21-8f6ffe9c34a3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ1NjUzODYsIm5iZiI6MTc0NDU2NTA4NiwicGF0aCI6Ii82MDA1NzMvNDMyMDQxMjQ4LTJlNzBmMmMyLTQ5ZWEtNDNiYS04YjIxLThmNmZmZTljMzRhMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxM1QxNzI0NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMDU1MDBhNWY5ZjcyNmNiYzY5OWM1YzlmYjI0YWI0ZjY4YWM0NzkyOTVmZjEzMGMwNjI5MzNkYjhkMTg1NzNhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.N0b3VuMr8mPkAdX8NaAmVt1-9eLYl58g6iblElQk4u4">

On my PC it I had a different issue, the marker was not covering the bottom line, while on your screenshot it's the top line. Could it be something related to inverted Y-axis on some renderers ? Anyway, I added +1 to the end of the line Y since `sstep = sty + 1`, could you confirm this works on your end ?

Lectem commented:
> It seems that there are subtle differences between sched_wakeup and sched_waking (there's also sched_wakeup_new) and that the origin CPU might not be the one of the wakee depending on various things ...
> 
> See https://perfetto.dev/docs/data-sources/cpu-scheduling
> 
> I'll investigate a bit more.
> 
> Note that on windows, it is possible that a thread seemingly unrelated readies (wakes up) another if it triggers the scheduler by calling any kind of synchronization or syscall. (this mostly happens when thread was already ready to be scheduled). From what I saw on windows, we're now displaying the same things as WPA and Pix.

In the end it seems that `sched_waking` is the event we needed, not `sched_wakeup`. 

`sched_wakeup` is only useful if you need to compute the scheduling latency of the context switch itself (ie: how long it takes for the thread to wakeup since the kernel actually decided to do it). Note that all `sched_wakeup` events are always followed by a `sched_switch` event, so it's not necessary to keep it as we do not compute the scheduling latency. 

Instead, I replaced `sched_wakeup` by `sched_waking`, which does give the information about the thread actually triggering the wakeup. This way now have the same information as the `ReadyThread` event on Windows. The only impact this has on current behavior is that the "Thread is waking up" part on the thread timeline will now indicate the same thing on both platforms, while it used to show only the scheduling latency on Linux.
![image](https://github.com/user-attachments/assets/60f7e0b2-fb3e-46d6-b584-46c40b1bc446) 

This was confirmed by analysing the ftrace output (for example here using `traceshark` (https://github.com/cunctator/traceshark).
![image](https://github.com/user-attachments/assets/5ba2f762-c747-4dbd-82ec-7d19b60260fc) )

Lectem commented:
> > The marker has incorrect offset.
> > <img alt="image" width="100" src="https://private-user-images.githubusercontent.com/600573/432041248-2e70f2c2-49ea-43ba-8b21-8f6ffe9c34a3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ1NjUzODYsIm5iZiI6MTc0NDU2NTA4NiwicGF0aCI6Ii82MDA1NzMvNDMyMDQxMjQ4LTJlNzBmMmMyLTQ5ZWEtNDNiYS04YjIxLThmNmZmZTljMzRhMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxM1QxNzI0NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMDU1MDBhNWY5ZjcyNmNiYzY5OWM1YzlmYjI0YWI0ZjY4YWM0NzkyOTVmZjEzMGMwNjI5MzNkYjhkMTg1NzNhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.N0b3VuMr8mPkAdX8NaAmVt1-9eLYl58g6iblElQk4u4">
> 
> On my PC it I had a different issue, the marker was not covering the bottom line, while on your screenshot it's the top line. Could it be something related to inverted Y-axis on some renderers ? Anyway, I added +1 to the end of the line Y since `sstep = sty + 1`, could you confirm this works on your end ?

So this was actually a half pixel issue.

I have a setup with the following screens:

- 2560x1440(1x) (main) + 2560x1600(1.5x) + 1080p (1x)
  When using this setup, whatever screen the application opens up on first, I had the following result (same as you):
  ![image](https://github.com/user-attachments/assets/faaaca63-005f-41bb-96d2-11b439e06c98)

- If I remove the first screen, and thus use the following setup:
   2560x1600(1.5x) (main) + 1080p (1x)
   ![image](https://github.com/user-attachments/assets/7ba440db-4826-45b5-9eb1-787989be37be)

- Then if I use only the 1080p (1x) screen, issue comes back.
   ![image](https://github.com/user-attachments/assets/f319200f-6f30-4434-a2f0-8fc386ecd820).


# Using a half pixel offset:

- 2560x1440(1x) (main) + 2560x1600(1.5x) + 1080p (1x)
   ![image](https://github.com/user-attachments/assets/484f989a-208f-48c8-9d2e-809590c8fbda)

-  2560x1600(1.5x) (main) + 1080p (1x)
  ![image](https://github.com/user-attachments/assets/bc2c5816-edd0-4702-ab24-521ad1f6bfc0)

- 1080p only:
   ![image](https://github.com/user-attachments/assets/08e97537-226e-4924-8be9-551f6042a20a)

This was also tested with different zoom levels and seems correct now.

wolfpld commented:
```
/home/wolf/tracy/profiler/src/profiler/TracyView_ContextSwitch.cpp:79:53: warning: expression result unused [-Wunused-value]
   79 |     case ContextSwitchData::Win32_WrPoolAllocation: "(Thread is waiting for a system allocation)";
      |                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

wolfpld commented on server/TracyWorker.cpp:
```diff
@@ -1419,7 +1419,8 @@ Worker::Worker( FileRead& f, EventType::Type eventMask, bool bgTasks, bool allow
     s_loadProgress.subTotal.store( 0, std::memory_order_relaxed );
     s_loadProgress.progress.store( LoadProgress::ContextSwitches, std::memory_order_relaxed );
-
+    
```
Spaces are added here.

Lectem replied:
Fixed, weirdly neither Github nor my diff tool showed those

Lectem commented:
I think with this last fix there should not be any pending issues anymore (as far as I'm concerned).
I've been using this for a while now (both on Windows and Linux), and didn't encounter other issues (not saying there aren't, but I've not seen them if there are).

Do you need me to address anything else for this PR ?

wolfpld commented:
I'll try to take a look.

gglin001 commented:
can we add support wakeup vis for `tracy-import-chrome` ?
